### PR TITLE
cover and toc option for wkhtmltopdf engine

### PR DESCRIPTION
**Pull request by TrainZug**
The toc parameter has no '--' or option. The cover parameter has no '--'.

### CODE REVIEW

**Comment by ADmad** [comment]
Please rebase your branch onto latest upstream `master` and add tests for your update.

**Comment by TrainZug** [comment]
I hope what I just did was ok. I'm still learning to use git

File: tests/TestCase/Pdf/Engine/WkHtmlToPdfEngineTest.php [hunk 1]
```diff
@@ -75,6 +75,19 @@ public function testGetCommand()
         $result = $method->invokeArgs($Pdf->engine(), []);
         $expected = "wkhtmltopdf --quiet --print-media-type --orientation 'portrait' --page-size 'A4' --encoding 'UTF-8' --boolean --string 'value' --integer '42' --array 'first' 'firstValue' --array 'second' 'secondValue' - -";
         $this->assertEquals($expected, $result);
+
+        $Pdf = new CakePdf([
+            'engine' => [
+                'className' => 'CakePdf.WkHtmlToPdf',
+                'options' => [
+                    'cover' => 'cover.html',
+                    'toc' => true,
```
Comment (ADmad) [inline tests/TestCase/Pdf/Engine/WkHtmlToPdfEngineTest.php:84]: From the `wkhtmltopdf`'s help the value for `toc` would be a string and you have instead only considered the case where it's value is `true`.

```
toc [TOC OPTION]...
  All options that can be specified for a page object can also be specified for
  a toc, further more the options from the TOC Options section can also be
  applied.
```

**Review by ADmad — COMMENTED** [review]
> I hope what I just did was ok. I'm still learning to use git

Exactly what I wanted :+1:.

File: tests/TestCase/Pdf/Engine/WkHtmlToPdfEngineTest.php (hunk 1 above)
Comment (TrainZug) [reply tests/TestCase/Pdf/Engine/WkHtmlToPdfEngineTest.php:84]: You can now define options for the toc and cover page via an array. I couldn't come up with a better solution to pass the url for the cover. So it has to be passed via the `url` index.

File: src/Pdf/Engine/WkHtmlToPdfEngine.php [hunk 2]
```diff
@@ -164,6 +157,54 @@ protected function _getCommand(): string
...
+                    throw new Exception('The url for the cover is missing. Use the "url" index.');
+                }
+                $command .= ' cover ' . escapeshellarg((string)$value['url']);
+                unset($value['url']);
+                foreach ($value as $k => $v) {
+                    $command .= $this->parseOptions($k, $v);
+                }
+            } else {
+                foreach ($value as $k => $v) {
+                    $command .= sprintf(' --%s %s %s', $key, escapeshellarg($k), escapeshellarg((string)$v));
```
[the remark this answers is not in the document]
Comment (TrainZug) [reply src/Pdf/Engine/WkHtmlToPdfEngine.php:188]: Yes, I implemented and tested that for the toc and cover options. You can see that here: [WkHtmlToPdfEngineTest.php](https://github.com/FriendsOfCake/CakePdf/pull/305/files/727862944741311c9e5c70f8e4ba13e5313d33e4#diff-9f11730ca67ba2c644fd23c99278db774ac09745d3e90b43f90d9ba52fdfdff9R132-R142)

File: src/Pdf/Engine/WkHtmlToPdfEngine.php [hunk 3]
```diff
@@ -164,6 +157,54 @@ protected function _getCommand(): string
         return $command;
     }
 
+    /**
+     * Parses a value of options to create a part of the command.
+     * Created to reuse logic to parse the cover and toc options.
+     *
+     * @param string $key the option key name
+     * @param string|bool|array $value the option value
```
Comment (ADmad) [inline src/Pdf/Engine/WkHtmlToPdfEngine.php:165]:
```suggestion
     * @param string|true|array $value the option value
```
Let's be extra accurate here :)

File: src/Pdf/Engine/WkHtmlToPdfEngine.php [hunk 4]
```diff
@@ -164,6 +157,54 @@ protected function _getCommand(): string
...
+            }
+        } elseif ($value === true) {
+            if ($key === 'toc') {
+                $command .= ' toc';
+            } else {
+                $command .= ' --' . $key;
+            }
+        } else {
+            if ($key === 'cover') {
+                $command .= ' cover ' . escapeshellarg((string)$value);
```
Comment (ADmad) [inline src/Pdf/Engine/WkHtmlToPdfEngine.php:199]:
```suggestion
                $command .= ' cover ' . escapeshellarg($value);
```

File: src/Pdf/Engine/WkHtmlToPdfEngine.php [hunk 5]
```diff
@@ -164,6 +157,54 @@ protected function _getCommand(): string
...
+            } else {
+                $command .= sprintf(' --%s %s', $key, escapeshellarg((string)$value));
```
Comment (ADmad) [inline src/Pdf/Engine/WkHtmlToPdfEngine.php:201]:
```suggestion
                $command .= sprintf(' --%s %s', $key, escapeshellarg($value));
```

**Comment by ADmad** [comment]
Nice work @TrainZug

**Pull request merged**